### PR TITLE
Benchmark improvements

### PR DESCRIPTION
**Pull request by kripken**
A bunch of benchmark improvements and cleanups, including

 * Move memcpy/memset and function pointer tests out of the core benchmarks (i.e., don't run them by default - they are slow and more detailed than regular use needs).
 * Update the flags we use when benchmarking (a few there were obsolete).
 * Add support for benchmarking [cheerp](https://www.leaningtech.com/cheerp/), which just launched wasm output. It's a little tricky (cheerp doens't use `main(argc, argv)`, so we need to rewrite the benchmark to use their `webMain()`), and many tests fail because of cheerp's limitations on things like `unions` or pointer casts or missing features, but most of the smaller one-file tests run with this PR. Nothing super interesting in the results: our wasm is smaller, their js is smaller, we are 12% faster on average.

### CODE REVIEW

File: tests/test_benchmark.py [hunk 1]
```diff
@@ -120,20 +124,153 @@ def process(filename):
...
+    })
+    cheerp_args = [
+      '-target', 'cheerp', '-cheerp-mode=wasm',
+    ]
+    self.parent = parent
+    if lib_builder:
+      # build as "native" (so no emcc env stuff), but with all the cheerp stuff
+      # set in the env
+      cheerp_args = cheerp_args + lib_builder(self.name, native=True, env_init={
+        'CC': '/opt/cheerp/bin/clang',
```
Comment (juj) [inline tests/test_benchmark.py]: This assumes Linux and that one has manually installed cheerp in advance? Btw, did cheerp build their own wasm backend or do they depend on upstream LLVM backend or JSBackend?

File: tests/test_benchmark.py [hunk 2]
```diff
@@ -120,20 +124,153 @@ def process(filename):
...
+        cheerp_temp,
+        '-Wno-writable-strings', # for how we set up webMain
+        '-o', final + '.wasm'
+      ] + shared_args
+      #print(' '.join(cmd))
+      subprocess.check_call(cmd)
+      self.filename = final
+      if self.binaryen_opts:
+        subprocess.check_call([
+          '/home/alon/Dev/2-binaryen/bin/wasm-opt',
```
Comment (juj) [inline tests/test_benchmark.py]: Here's a hardcoded path to your own wasm installation.

File: tests/test_benchmark.py (hunk 1 above)
Comment (kripken) [reply tests/test_benchmark.py]: Yeah, that's how cheerp installs itself on linux, always to that path. Not sure how to get this to work on other platforms.

Cheerp built their [own wasm writing backend](https://github.com/leaningtech/cheerp-llvm/blob/master/lib/CheerpWriter/CheerpWasmWriter.cpp). Mostly straightfoward writing, little wasm-specific optimizations so far (the binaryen optimizer shrinks their binaries by 15%, I mentioned it to them).

File: tests/test_benchmark.py (hunk 2 above)
Comment (kripken) [reply tests/test_benchmark.py]: Yeah, we don't have a good way to do this actually. I guess now is as good a time as any to fix that :)

Comment (kripken) [reply tests/test_benchmark.py]: Done.

**Comment by kripken** [comment]
This uncovered a bug with the wasm backend not building the binaryen port properly. Added a fix - we need to build it only when we use it, not early in emcc (as we reach that when building libc etc., so we'd infinitely recurse). We had this bug before actually, but looks like we never added a test - added one now.

File: tests/test_benchmark.py (hunk 1 above)
Comment (juj) [reply tests/test_benchmark.py]: Is it possible to add a `if (not exists /opt/cheerp/bin/clang) self.skip('missing cheerp')` or similar? I think that's what we added for Crunch and SCons and some others in the suites.

**Comment by kripken** [comment]
Any objections to landing this? It's mostly just tweaks for benchmarking, but the refactor and fix for building binaryen with the wasm backend is important.

File: tests/test_benchmark.py (hunk 1 above)
Comment (kripken) [reply tests/test_benchmark.py]: sure, adding

**Comment by kripken** [comment]
Merging: this fixes a critical wasm-backend bug, @juj's review comments should all be addressed, and he's gone on vacation.

**Pull request merged**